vllm/distributed/device_communicators/custom_all_reduce_utils.py

```diff
@@ -0,0 +1,165 @@
+import json
+import os
+import sys
+from contextlib import contextmanager
+from typing import Callable, Dict, List, Optional
+
+import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+
+import vllm.envs as envs
+from vllm.distributed.parallel_state import get_cpu_world_group, get_local_rank
+from vllm.logger import init_logger
+
+logger = init_logger(__name__)
+
+
+@contextmanager
+def mute_output():
+    with open(os.devnull, "w") as f:
+        sys.stderr = f
+        sys.stdout = f
+        yield
+
+
+def producer(i, CUDA_VISIBLE_DEVICES=None):
```

Comment:
Please annotate the type
```suggestion
def producer(i, CUDA_VISIBLE_DEVICES: Optional[str] = None):
```

Reply:
Also, it looks a bit bizarre to me that the parameter name uses the uppercase letters. Why not `cuda_visible_devices` or something like that?